src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+        pm::HSBClock::FLASHCALWP, pm::PBBClock::FLASHCALW);
+
+
+// Few constants relating to module configuration.
+const FLASH_PAGE_SIZE : u32 = 512;
+const FLASH_NB_OF_REGIONS : u32 = 16;
+const FLASHCALW_REGIONS : u32 = FLASH_NB_OF_REGIONS;
+const FLASHCALW_CMD_KEY : u32 = 0xA5 << 24;
+
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS1_FWS_1_FWU_MAX_FREQ : u32 = 12000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS0_FWS_0_MAX_FREQ : u32 = 18000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS0_FWS_1_MAX_FREQ : u32 = 36000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS1_FWS_0_MAX_FREQ : u32 = 8000000;
+
+
+#[cfg(not(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE))]
+const FLASH_FREQ_PS2_FWS_0_MAX_FREQ : u32 = 24000000;
+
+const GP_ALL_FUSES_ONE : u64 = !0 as u64;
+
+// Macros for getting the i-th bit.
+macro_rules! get_bit {
+    ($w:expr) => (0x1u32 << $w);
+}
+
+/// This is simply std::cmp::min from std
+#[inline]
+fn min<T: Ord>(v1: T, v2: T) -> T {
+    if v1 <= v2 { v1 } else { v2 }    
+}
+
+// Should never ever have to wait (because we're using interrupts!)
+pub fn default_wait_until_ready(flash : &FLASHCALW) {
+    while !flash.get_ready_status() { 
+        panic!("Flash wasn't ready for a command -- interrupt system used incorrectly.");
+    }
+}
+
+/// Trait for a client of the flash controller.
+pub trait Client {
+    //  Called upon a completed call
+    fn command_complete(&self, err: Error);     
+}
+    
+impl FLASHCALW {
+
+    pub fn mark_ready(&self) {
+        self.ready.put(Some(true));
+    }
+
+
+    pub fn get_ready_status(&self) -> bool {
+        if self.ready.is_none() || !self.ready.take().unwrap() {
+            self.ready.put(Some(false));
+            false
+        } else {
+            self.ready.put(Some(true));
+            true
+        }
+    }
+    
+    const fn new(base_addr: *mut Registers, ahb_clk: pm::HSBClock,
+    hramc1_clk: pm::HSBClock, pb_clk: pm::PBBClock) -> FLASHCALW {
+        FLASHCALW {
+            registers:              base_addr,
+            ahb_clock:              pm::Clock::HSB(ahb_clk),
+            hramc1_clock:           pm::Clock::HSB(hramc1_clk),
+            pb_clock:               pm::Clock::PBB(pb_clk),
+            wait_until_ready:       default_wait_until_ready,
+            error_status:           TakeCell::new(0),
+            ready:                  TakeCell::new(true),
+            client:                 TakeCell::empty(),
+            current_state:          Cell::new(FlashState::Unconfigured),
+            current_command:        Cell::new(Command::None),
+            page:                   Cell::new(0),
+            page_buffer:            TakeCell::new([0; FLASH_PAGE_SIZE as usize])
+        }
+    }
+
+    // Helper to read a flashcalw register 
+    fn read_register(&self, key : RegKey) -> u32 {
```

Comment:
code calling this should just directly dereference the registers themselves, it's much clearer and almost certainly way more efficient, unless the compiler does some serious in-lining (which it might)